[BUG] go-api crashes from malformed PDF panic — HN pipeline fetch step fails with exit 22

## 🐛 Problem
- The `Fetch articles` step of the HN pipeline (`.github/workflows/hackernews-pipeline.yml`) fails with curl exit code 22.
- Root cause: while scraping a malformed PDF, `rsc.io/pdf` throws a `panic` instead of returning an `error`. The worker goroutine has no `recover`, so the panic propagates up and **kills the entire Go process**. Since `compose.prod.yml` defines no `restart:` policy for `go-api`, the container stays **dead with no auto-recovery**. Every subsequent request then hits a 502 from the upstream proxy, which causes `curl -sf` to exit 22.
- Re-running the workflow reproduces the same failure because `go-api` is still down.

## 🔁 Steps to reproduce
1. Trigger the HN pipeline via `workflow_dispatch` with `date=260416` (or any date that includes the PDF below in the HN front page).
2. `POST /hackernews/fetch/260416` returns `200` and enqueues 95 tasks.
3. The scraper picks up `https://fingfx.thomsonreuters.com/gfx/legaldocs/xmvjyjekkpr/Rakoff%20-%20order%20-%20AI.pdf`.
4. `p.Content()` at `services/go/internal/scraper/pdf.go:70` panics.
5. Subsequent `GET /hackernews/fetch/status/260416` calls get 502 from the proxy → workflow step exits 22.

## 🎯 Expected
- A single broken PDF should fail only that one task; the other 94 should continue.
- The process should stay alive; even if it dies, the container must auto-restart.
- The whole pipeline must not be taken down by a single bad URL.

## ❌ Actual
- Process-wide crash, no container restart, pipeline fails, go-api remains down until manual intervention.

```
2026/04/15 19:54:10 [scraper] PDF detected: url=https://fingfx.thomsonreuters.com/gfx/legaldocs/xmvjyjekkpr/Rakoff%20-%20order%20-%20AI.pdf
...
panic: malformed PDF: reading at offset 0: stream not present

goroutine 75 [running]:
rsc.io/pdf.(*buffer).errorf(...)
        /go/pkg/mod/rsc.io/pdf@v0.1.1/lex.go:82 +0x48
rsc.io/pdf.(*buffer).reload(...)
        /go/pkg/mod/rsc.io/pdf@v0.1.1/lex.go:95 +0x145
rsc.io/pdf.Page.Content(...)
        /go/pkg/mod/rsc.io/pdf@v0.1.1/page.go:440 +0x2ca
blog-go-api/internal/scraper.extractPDFText(...)
        /app/internal/scraper/pdf.go:70 +0x12a
blog-go-api/internal/scraper.FetchPDFContent(...)
        /app/internal/scraper/pdf.go:47 +0x511
blog-go-api/internal/scraper.FetchContent(...)
        /app/internal/scraper/content.go:131 +0x2fe
main.main.FetchHandler.func4.1()
        /app/internal/handler/fetch.go:88 +0x11a
blog-go-api/internal/worker.(*Pool).Submit.func1()
        /app/internal/worker/pool.go:26 +0x6e
```

## 🖥 Environment
- Prod EC2, `ghcr.io/foreverfl/blog-v2/go-api:v0.2.10` (debian:bookworm-slim base)
- Go 1.26, `rsc.io/pdf v0.1.1`
- Deployed via `infra/docker/compose.prod.yml`

## 🔍 Root cause (three layers)
1. **L1 — upstream library defect**: `rsc.io/pdf@v0.1.1` panics on malformed PDF streams instead of returning an error. The library is a Russ Cox experimental module with no robustness guarantees; this class of panic is known upstream and effectively unmaintained.
2. **L2 — missing panic boundary in scraper**: `scraper/pdf.go:extractPDFText` calls `p.Content()` with no `recover`, so the L1 panic propagates unchanged to the caller.
3. **L3 — blast radius amplification**: `worker/pool.go:Submit` spawns goroutines without `recover`, so a single task panic kills the entire process. `compose.prod.yml` sets no `restart:` policy for `go-api` (nor for `rust-api`/`auth-api`/`haskell-api`), so the dead container is never auto-recovered. A per-URL failure escalates into a full-service outage.

Note: this is **unrelated** to the logging middleware work in `fe13f81`..`7629e1f`. That series only touches `internal/middleware/logging.go`; neither `scraper/pdf.go` nor `worker/pool.go` was modified. The panic trigger is new input data (this specific PDF first appearing on HN on 260416), not a code regression. The latent bomb has been in the code since `b6dedb5`.

## ✅ Fix criteria
- [ ] Reproduced locally by pointing the scraper at the Rakoff PDF URL
- [ ] L2 fix: `defer recover()` in `scraper/pdf.go:extractPDFText` converts panic → error so the task fails gracefully
- [ ] L3 fix (defense in depth): `defer recover()` in `worker/pool.go:Submit` so any future panic in any task cannot take down the process; log the recovered value
- [ ] L3 fix (infra): add `restart: unless-stopped` to `go-api`, `rust-api`, `auth-api`, `haskell-api` in `compose.prod.yml` — covers SIGSEGV / OOM kill that `recover` cannot catch
- [ ] Prod go-api brought back up (`docker compose -f docker-compose.prod.yml up -d go-api`) and HN pipeline re-run for 260416 to confirm recovery
- [ ] Follow-up (optional): evaluate replacing `rsc.io/pdf` with a more robust parser (e.g. `github.com/ledongthuc/pdf`, `github.com/dslipak/pdf`) or guard PDF fetches behind a subprocess so a crash is isolated

--- a/infra/docker/compose.prod.yml
+++ b/infra/docker/compose.prod.yml
@@ -21,6 +21,7 @@ services:
     ports:
       - "8001:8001"
     env_file: .env.prod
+    restart: unless-stopped
     depends_on:
       flyway:
         condition: service_completed_successfully
@@ -32,6 +33,7 @@ services:
     ports:
       - "8002:8002"
     env_file: .env.prod
+    restart: unless-stopped
     depends_on:
       flyway:
         condition: service_completed_successfully
@@ -43,6 +45,7 @@ services:
     ports:
       - "8003:8003"
     env_file: .env.prod
+    restart: unless-stopped
     depends_on:
       flyway:
         condition: service_completed_successfully
@@ -54,6 +57,7 @@ services:
     ports:
       - "8004:8004"
     env_file: .env.prod
+    restart: unless-stopped
     depends_on:
       flyway:
         condition: service_completed_successfully

--- a/services/go/internal/scraper/pdf.go
+++ b/services/go/internal/scraper/pdf.go
@@ -53,7 +53,15 @@ func FetchPDFContent(url string) (string, error) {
 	return text, nil
 }
 
-func extractPDFText(path string) (string, error) {
+func extractPDFText(path string) (text string, err error) {
+	// rsc.io/pdf panics on malformed PDF streams instead of returning an error.
+	// Contain it here so one bad PDF fails only this task, not the whole process.
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("pdf parse panic: %v", r)
+		}
+	}()
+
 	r, err := pdf.Open(path)
 	if err != nil {
 		return "", fmt.Errorf("open pdf: %w", err)
@@ -75,7 +83,7 @@ func extractPDFText(path string) (string, error) {
 		buf.WriteString("\n\n")
 	}
 
-	text := buf.String()
+	text = buf.String()
 	text = pdfTrailingWs.ReplaceAllString(text, "\n")
 	text = pdfMultiNewline.ReplaceAllString(text, "\n\n")
 	return strings.TrimSpace(text), nil

--- a/services/go/internal/worker/pool.go
+++ b/services/go/internal/worker/pool.go
@@ -1,6 +1,10 @@
 package worker
 
-import "sync"
+import (
+	"log"
+	"runtime/debug"
+	"sync"
+)
 
 // Pool provides bounded concurrency using a semaphore pattern.
 type Pool struct {
@@ -20,6 +24,9 @@ func (p *Pool) Submit(fn func()) {
 	go func() {
 		p.sem <- struct{}{}
 		defer func() {
+			if r := recover(); r != nil {
+				log.Printf("[worker] task panic: %v\n%s", r, debug.Stack())
+			}
 			<-p.sem
 			p.wg.Done()
 		}()

